### DISCUSSION TITLE
CSVWriter using multiple inputs

### DISCUSSION THREAD

**Discussion by seiedsalad** [bonsai-rx/bonsai#1771] [Q&A]
I'm going to be getting parallel outputs from multiple RFID readers as well as an Arduino setup over 30 minutes in each session that will all be sending data to their respective com ports, which I have already set up with SerialStringRead and subsequent Timestamp nodes for each com port. However, I want to get around the problem of getting n# CSV files as a result and needing to manually combine all of these outputs into one table and sort out the timestamps myself. I'm not able to just connect a single CSV Writer node to every one of these individual paths because the node will not take inputs from multiple sources. I've also tried connecting all of these com port paths to a zip node and connecting that to a csvwriter node, but that will only result in the final CSV giving me one string value and one timestamp item from each port. I'm including pictures of both my workflow right now, as well as the 2 separate test csv files that result from this. the left file shows the serial monitor output from a simple blink sketch on my arduino, and the right shows the rfid reader output (there will be more than 10 rfid readers in total but this is just the setup I'm using to test it out).
![Workflow](https://github.com/bonsai-rx/bonsai/assets/168840703/67a00d54-1bdd-49e1-99bc-8ad992001774)
![csv data](https://github.com/bonsai-rx/bonsai/assets/168840703/1de7592b-5c75-40b9-a4e3-b219ad0ee040)

**Comment by bruno-f-cruz** [comment]
You can use a `merge` operator:
![image](https://github.com/bonsai-rx/bonsai/assets/7049351/b61fe912-220b-4afb-8e18-a4acd69fbb86)

or, even better, multicast all the streams to the same Subject Source:

![image](https://github.com/bonsai-rx/bonsai/assets/7049351/6dd73109-f8fc-4a8f-b567-2cff7323d0b6)

**Reply by seiedsalad** [reply]
Thanks for the reply. What would be the best subject source to use in this instance?

**Reply by bruno-f-cruz** [reply]
Probably a BehaviorSubject.

**Reply by seiedsalad** [reply]
I'm getting this error when trying to implement the BehaviorSubject. Does this setup look right? Apologies for the confusion - I'm very new to Bonsai!
![BehaviorSubject](https://github.com/bonsai-rx/bonsai/assets/168840703/11ecda17-3e25-42f0-ab69-3dcb65ddee28)

**Reply by bruno-f-cruz** [reply]
Try following the docs instructions:

https://bonsai-rx.org/docs/articles/subjects.html#source-subjects

**Reply by seiedsalad** [reply]
I've resorted to just using a Merge operator which seems to be working well. However, now I'm running into the issue of what information is being written onto the csv file. Currently, all that is being printed as a value for each read of the serial monitor is "\r\n\". For each com port, I need to know what exactly is being read on the serial monitor (i.e., the rfid tag number) as well as the timestamp of that output, and which port said output is coming from. Any ideas what could be going wrong? the problem is only occurring after I externalize the properties of the SerialStringRead node, but I believe that is a necessary step in order to know which port any given value is coming from on the csv file. I've attached pictures of both my workflow and an example csv file as a result.
![workflow (1)](https://github.com/bonsai-rx/bonsai/assets/168840703/3c77eda1-54c2-448a-9d72-b226a66780ee)
![csv file](https://github.com/bonsai-rx/bonsai/assets/168840703/54e24ff0-3230-4bab-bd9e-8a761dd46f6f)
